systems/plants/@RigidBodyManipulator/manipulatorDynamics.m

```diff
@@ -1,221 +1,515 @@
+          H(n,n) = S{i}' * fh;
+          dH(n + (n-1)*m.NB,nk) = S{i}' * dfh;
+          j = i;
+          while m.parent(j) > 0
+            if j==k,
+              dfh = Xup{j}' * dfh + dXupdq{j}' * fh;
+            else
+              dfh = Xup{j}' * dfh;
+            end
+            fh = Xup{j}' * fh;
+            
+            j = m.parent(j);
+            np = m.position_num(j);
+            
+            H(n,np) = S{j}' * fh;
+            H(np,n) = H(n,np);
+            dH(n + (np-1)*m.NB,nk) = S{j}' * dfh;
+            dH(np + (n-1)*m.NB,nk) = dH(n + (np-1)*m.NB,nk);
+          end
         end
-        dIC{m.parent(i),n} = dIC{m.parent(i),n} + ...
-          dXupdq{i}'*IC{i}*Xup{i} + Xup{i}'*IC{i}*dXupdq{i};
       end
+      
+      dH = dH(:,1:m.NB)*[eye(m.NB) zeros(m.NB)];
+      dC(:,m.NB+1:end) = dC(:,m.NB+1:end) + diag(m.damping);
+      
+      ind = find(abs(qd)<m.coulomb_window');
+      dind = sign(qd(ind))./m.coulomb_window(ind)' .* m.coulomb_friction(ind)';
+      fc_drv = zeros(m.NB,1);
+      fc_drv(ind) =dind;
+      dC(:,m.NB+1:end) = dC(:,m.NB+1:end)+ diag(fc_drv);
+    else
+      [H,C] = HandC(m,q,qd,f_ext,obj.gravity);
     end
     
-    % minor adjustment to make TaylorVar work better.
-    %H = zeros(m.NB);
-    H=zeros(m.NB)*q(1);
-    
-    %Derivatives wrt q(k)
-    dH = zeros(m.NB^2,2*m.NB)*q(1);
-    for k = 1:m.NB
-      nk = m.position_num(k);
-      for i = 1:m.NB
-        n = m.position_num(i);
-        fh = IC{i} * S{i};
-        dfh = dIC{i,nk} * S{i};  %dfh/dqk
-        H(n,n) = S{i}' * fh;
-        dH(n + (n-1)*m.NB,nk) = S{i}' * dfh;
-        j = i;
-        while m.parent(j) > 0
-          if j==k,
-            dfh = Xup{j}' * dfh + dXupdq{j}' * fh;
-          else
-            dfh = Xup{j}' * dfh;
-          end
-          fh = Xup{j}' * fh;
-          
-          j = m.parent(j);
-          np = m.position_num(j);
-          
-          H(n,np) = S{j}' * fh;
-          H(np,n) = H(n,np);
-          dH(n + (np-1)*m.NB,nk) = S{j}' * dfh;
-          dH(np + (n-1)*m.NB,nk) = dH(n + (np-1)*m.NB,nk);
-        end
+    C=C + computeFrictionForce(obj,qd);
+  end
+end
+
+end
+
+function [H,C,B,dH,dC,dB] = manipulatorDynamicsNew(obj,q,v,use_mex)
+% manipulatorDynamics  Calculate coefficients of equation of motion.
+% [H,C,B,dH,dC,dB] = manipulatorDynamics(obj,q,v,use_mex) calculates the
+% coefficients of the joint-space equation of motion,
+% H(q)*vd+C(q,v,f_ext)=B(q,v)*tau, where q, v and vd are the joint
+% position, velocity and acceleration vectors, H is the joint-space inertia
+% matrix, C is the vector of gravity, external-force and velocity-product
+% terms, and tau is the joint force vector.
+%
+% Algorithm: recursive Newton-Euler for C, and Composite-Rigid-Body for H.
+
+compute_gradients = nargout > 3;
+
+options.use_mex = use_mex;
+options.compute_gradients = compute_gradients;
+options.compute_JdotV = true;
+kinsol = doKinematics(obj, q, v, options);
```

Comment:
so right now we are relying on the caching mechanism here.  should we change manipulatorDynamics to take a kinsol as an input argument?  it would be more flexible (e.g. for the trajectory optimization which might call things out of order)